Track object modifications in fake clientset

olegshaldybin opened the pull request:
Fake clientset is used by unit tests extensively but it has some
shortcomings:
- no filtering on namespace and name: tests that want to test objects in
  multiple namespaces end up getting all objects from this clientset,
  as it doesn't perform any filtering based on name and namespace;
- updates and deletes don't modify the clientset state, so some tests
  can get unexpected results if they modify/delete objects using the
  clientset;
- it's possible to insert multiple objects with the same
  kind/name/namespace, this leads to confusing behavior, as retrieval is
  based on the insertion order, but anchors on the last added object as
  long as no more objects are added.

This change changes core.ObjectRetriever implementation to track object
adds, updates and deletes.

Some unit tests were depending on the previous (and somewhat incorrect)
behavior. These are fixed in the following few commits.

olegshaldybin committed:
Track object modifications in fake clientset

Fake clientset is used by unit tests extensively but it has some
shortcomings:

- no filtering on namespace and name: tests that want to test objects in
  multiple namespaces end up getting all objects from this clientset,
  as it doesn't perform any filtering based on name and namespace;

- updates and deletes don't modify the clientset state, so some tests
  can get unexpected results if they modify/delete objects using the
  clientset;

- it's possible to insert multiple objects with the same
  kind/name/namespace, this leads to confusing behavior, as retrieval is
  based on the insertion order, but anchors on the last added object as
  long as no more objects are added.

This change changes core.ObjectRetriever implementation to track object
adds, updates and deletes.

Some unit tests were depending on the previous (and somewhat incorrect)
behavior. These are fixed in the following few commits.

olegshaldybin committed:
Fix expectations in deployment controller test

Since fake clientset now correctly tracks objects created by deployment
controller, it triggers different controller behavior: controller only
creates replica set once and updates deployment once.

olegshaldybin committed:
Don't panic in NodeController if pod update fails

Previously it was trying to use a nil pod variable if error was returned
from the pod update call.

olegshaldybin committed:
Fix mirror pod identity change test

If mirror pod identity changes it actually doesn't get updated;
previously test was relying on an incorrect record pulled from a fake
clientset.

olegshaldybin committed:
Use correct namespace in unit tests that use fake clientset

Fake clientset no longer needs to be prepopulated with records: keeping
them in leads to the name conflict on creates. Also, since fake
clientset now respects namespaces, we need to correctly populate them.

olegshaldybin committed:
Expect namespace deletion in NamespaceController

Using new fake clientset registry exposes the actual flow on pending
namespace finalization: get namespace, create finalizer, list pods and
delete namespace if there are no pods.

olegshaldybin commented:
Any chance someone can take a look, or kick off tests for this branch? We are planning on another PR son, and it depends on the corrected behavior. Would be great to get this in if possible (or get someone's opinion on why it's not a good idea).

mikedanese commented:
@deads2k @caesarxuchao

caesarxuchao commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
-		switch castAction := action.(type) {
-		case ListAction:
+		// Here and below we need to switch on implementation types,
+		// no on interfaces, as some interfaces are identical
```
s/no/not

olegshaldybin commented:
Fixed a typo and rebased.

caesarxuchao commented:
@olegshaldybin Thanks for the PR. I agree with your first and third points, but I'm not sure about the second point (`updates and deletes don't modify the clientset state`). I think if a test needs to change the objects in the registry, then it should be an integration test that runs a real API server. 
IMO, the fake client should be dumb and should only return pre-programmed objects, and perhaps we should name it `mock` client rather than `fake`. @deads2k what do you think?

olegshaldybin commented:
@caesarxuchao That makes sense, however there are existing unit tests that rely on modifying objects using the clientset and observing the modified record later on: this can get confusing if they retrieve the old version of the resource after the update. If you want I can try and reduce this change to just handle namespace/name filtering and duplicate insertions, but that can lead to some tests not really testing the real flow (e.g. namespace finalizer test, deployment controller test).

caesarxuchao commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -778,18 +773,13 @@ func TestSyncDeploymentCreatesReplicaSet(t *testing.T) {
-	f.expectListPodAction(rs.Namespace, opt)
-	f.expectUpdateDeploymentAction(d)
+	f.expectListPodAction(rs.Namespace, api.ListOptions{})
 
```
This is a nice catch. For a moment I thought the old test was correct, but after some debugging I realized you are right. cc @janetkuo (for she maintains the deployment API).

This is a good example of why a minimum registry in the fake clientset is useful, because *i)*syncDeployment() has multiple interactions with the API server and its behavior is driven by the states stored in the API server, and *ii)* we want to verify the sequence of the client requests, which currently cannot be done in an integration test.

An alternative solution is running this as an integration test, and add a wrapper around a real clientset to record all the requests sent. cc @lavalamp.

janetkuo replied:
Thanks, this is correct. 
The controller first syncs the deployment and then creates an rs with 1 replica immediately (`expectCreateRSAction`), update the deployment's revision (`expectUpdateDeploymentAction`), and then list pods to calculate deployment status (`expectListPodAction`).

janetkuo commented:
ok to test

deads2k commented:
> I think if a test needs to change the objects in the registry, then it should be an integration test that runs a real API server.

I find @olegshaldybin's argument pretty compelling.  I'll review the pull today, but if the code looks reasonably simple I think I'd be in favor of allowing it.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+
 		kind, err := mapper.KindFor(resource)
+		if err != nil {
+			return false, nil, fmt.Errorf("unrecognized action for resource %s: %s", resource, err)
```
"no kind for registered for resource"?  I know its pre-existing, but you've already tweaked it some.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -18,28 +18,37 @@ package core
+	// kind is a list, all objects that match ns and name are returned
+	// in their corresponding list enclosure (e.g. PodList,
+	// ServiceList).
+	Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error)
```
Can we get a separate list method?

olegshaldybin replied:
Yeah, it will make things cleaner for sure. I am still not clear on a distinction between GroupResource and GroupKind: existing code seems to be using them interchangeably.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+			if typedAction.GetSubresource() == "" {
+				err = registry.Add(typedAction.GetObject())
+			} else {
+				// Handle subresource creation as an update on the
```
Can you model this as a separate method that takes the subresource and the object and must return the correct value to you?  Some subresource posts require special handling.  I can certainly see making a standard implementation for handling "status" since its so common.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+type registry struct {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
```
`lock`?

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
+	objects map[string][]runtime.Object
```
doc describing key/value.  I was expecting `map[string]map[string]runtime.Object` for namespace, name, object.

deads2k replied:
Actually, I think this should be 
`map[unversioned.GroupResource]map[string(namespace)]map[string(name)]runtime.Object`  I don't think you should index by kind, since we access by resource.

olegshaldybin replied:
Should it be `GroupResource` or `GroupVersionKind`? It seems natural to bucket them by `GroupVersionKind`, since that's what we can extract from the object being added. I'd like to understand more about the difference between `Resource` and `Kind`.

caesarxuchao replied:
I would say it should be `GroupVersionResource` here? Why `GroupResource`? 

> It seems natural to bucket them by GroupVersionKind, since that's what we can extract from the object being added

Actually I think we shouldn't extract it from the object. We should use the `GroupVersionResource` populated from the `Action`.

> I'd like to understand more about the difference between Resource and Kind

`GroupVersionResource` indicates what URL the request will be sent to. `GroupVersionKind` is a property of the object. In most cases they are the same, but sometimes GVK can be translated to different GVRs (e.g. `Scale`). Also, GVK.Version can be empty or "__internalVersion", because we have an internal version for objects, but GVR should never have an empty version, because there is always a version in the URL in k8s.

olegshaldybin replied:
Thanks, that's very helpful.

olegshaldybin replied:
@caesarxuchao Here's one thing I am not sure how to implement: say we want to have a `List()` method that receives a `GroupVersionResource` argument: now we want to create a list of objects of a certain kind. How do I translate the incoming `GroupVersionResource` to a `GroupVersionKind` of the correct `List` object?

olegshaldybin replied:
Ah, I think I can do it by using `mapper.KindFor`.

caesarxuchao replied:
@olegshaldybin could you point me to the code where you need to create the list of objects? I must miss something, because I think we don't need to get `GVK`. I hope we don't need to use `mapper` in the fake clientset.

olegshaldybin replied:
`GVK` is actually used all over the place in the existing implementation. Here's where it creates the  result object based on a `GVK` passed in:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L17

It's already using mapper to retrieve `GVK` from a `GVR`:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L61

I am definitely interested in improving some of these patterns, but not sure if we can avoid using `GVK`. To be fair, I don't have a full understanding of k8s API machinery yet, just started looking at how it's organized, so somewhat hesitant to change it too much. But if you have some pointers to a better approach, I can take a stab at it.

olegshaldybin replied:
Also, when setting tests up, objects are being added to the clientset directly, at that point we don't know what `GroupVersionResource` to associate them with, since we don't have any REST context at that point, just a `runtime.Object`.

caesarxuchao replied:
I haven't looked carefully, but I think we can get rid of GVK in most places, but we still need GVK where we called [o.scheme.New()](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L161) to get List objects and empty objects.

I'm keen on getting rid of calls to mapper.Kind() because that won't work once we convert controllers to use versioned clientsets, because mapper.Kind() always return the internal kind, while the controllers will expect to receive versioned Lists. Anyway, this is a separate issue, use GVK where you have to and we'll deal with it later.

caesarxuchao replied:
> Also, when setting tests up, objects are being added to the clientset directly, at that point we don't know what GroupVersionResource to associate them with, since we don't have any REST context at that point, just a runtime.Object.

Right. Perhaps we can ask the caller to provide GVR along with the objects, but that will be a separate PR.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+	}
+
+	nilValue := reflect.Zero(reflect.TypeOf(out)).Interface().(runtime.Object)
+	isList := strings.HasSuffix(kind.Kind, "List")
```
You should be able to do `meta.IsListType(scheme.New(kind))`

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
-func (o objects) Add(obj runtime.Object) error {
-	gvks, _, err := o.scheme.ObjectKinds(obj)
+func (r *registry) Add(obj runtime.Object) error {
+	return r.add(obj, false)
```
Check for existence before starting, you should produce an "already exists" error.

caesarxuchao replied:
I think the check is done in `add()`.

olegshaldybin replied:
add() already does that: if existing object matches kind, namespace and name of the one being inserted it returns an "already exists" error.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
+}
+
+func (r *registry) Update(obj runtime.Object) error {
+	return r.add(obj, true)
```
check for existence before starting, you should produce a "not found" error.

olegshaldybin replied:
add() does that as well, if replaceExisting flag is true but no existing object found it returns a "not found" error.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered object kind for %v", obj)
+	}
+	groupResource := unversioned.GroupResource{Group: gvks[0].Group, Resource: gvks[0].Kind}
```
You'll need to plumb the actual resource down.

deads2k commented:
I like it in concept, but the implementation needs to keep the distinction between kind and resource clear.  Looking at what's here, I think the "update" and "add" affecting "get" and "list" looks like a pretty natural outgrowth.

olegshaldybin commented:
@deads2k Thanks for you review! I was trying to minimize the actual code change for the additions but most of your feedback makes sense, I'll go ahead and incorporate it.

caesarxuchao commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
+
+// NewObjects returns a Registry that can be used to keep track of
+// objects for the fake clientset. Mostly useful for unit tests.
+func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) Registry {
```
Rename to NewRegistry?

olegshaldybin replied:
I would love to, but that will require regenerating a fake clientset code. Do you think it's worth it?

janetkuo commented:
@k8s-bot ok to test

caesarxuchao commented:
Could you add a comment to `NewSimpleClientset` to say something like "the returned fake clientset by default uses a fake registry, which may not reflect the real behavior of the API server (e.g., no validation or defaulting), and if the user wish to *not* use the fake registry, they can specify a different Reactor"?

olegshaldybin commented:
@caesarxuchao Sure, will do.

caesarxuchao commented:
`NewSimpleClientset` is [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/release_1_3/fake/clientset_generated.go#L39). That will also require regenerating the clientset.

The template is [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go#L115), you just need to change the template and then run hack/update-codegen.sh.

lavalamp commented:
I would kind of prefer it if you did not get a fake registry by default, if
you had to invoke something special to get it. This is because I don't
expect it to perfectly replicate a real registry (validation, etc) and I
don't want tests to unintentionally assume that it does.

On Tue, Jun 7, 2016 at 10:32 AM, Oleg Shaldybin <notifications@github.com>
wrote:

> @caesarxuchao <https://github.com/caesarxuchao> Sure, will do.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26680#issuecomment-224354614>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglu5DvJXZq58Y9KaAmwkMBfmyUufWks5qJashgaJpZM4IsMqJ>
> .
>

olegshaldybin commented:
@lavalamp But it seems to me even worse if tests assume that their previous modification actually went through because client returned no error, but getting the old version of the object on the next request, since the registry only contains objects explicitly added by the test framework.

lavalamp commented:
It depends on what sort of a test you're writing. If you want to verify an
exact sequence of calls (a form of test that IMO is typically bad outside
of special circumstances) then you should expect that behavior, it's how
all the mocking libraries I'm aware of work.

If you want to verify some other behavior, like a correct result after a
sequence of calls, then you probably want a real integration test, with a
real registry. Otherwise what are you really verifying?

On Tue, Jun 7, 2016 at 10:43 AM, Oleg Shaldybin <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> But it seems to me even worse if
> tests assume that their previous modification actually went through because
> client returned no error, but getting the old version of the object on the
> next request, since the registry only contains objects explicitly added by
> the test framework.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26680#issuecomment-224357893>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAnglgSDsnDX9Jt1eXTm-2EhI7_ZekDQks5qJa22gaJpZM4IsMqJ>
> .
>

olegshaldybin commented:
@lavalamp I understand your concern, but in that case only read operations should be allowed against this default reactor, and any modifications should return an error. Otherwise it's easy to write unit tests that don't reflect the actual system behavior (you can see some examples in this PR).

I can reduce the change to just apply namespace filtering properly if that's the consensus.

caesarxuchao commented:
> I would kind of prefer it if you did not get a fake registry by default, if you had to invoke something special to get it.

@olegshaldybin could you do this? Maybe adding a `bool fakeRegisty` flag to `NewSimpleClientset` and fall back to the old behavior if the flag is false (this could be messy..)? This will make sure people understand they are not interacting with a real registry.
@lavalamp do you think adding a comment stating "the fake registry is not an accurate replica of the real registry" to the `fake.NewSimpleClientset` is enough?

olegshaldybin commented:
@caesarxuchao Flag is doable but it can get pretty messy. I'll try and experiment with it though.

olegshaldybin committed:
Regenerate clientsets

olegshaldybin commented:
@caesarxuchao @deads2k Sorry, I just got back to this change today. Responded to most feedback and renamed Registry to ObjectTracker to avoid any confusion with real registries. Also clarified the intent in NewSimpleClientset.

Changing internal tracking to be based `GroupVersionResource` probably needs to be its own PR, as it's not straightforward and requires changes to all tests that use the clientset, as currently they don't supply the `GroupVersionResource`, just objects themselves. With the way it's currently used by tests it makes more sense to bucket them by `GVK`. Note that `GVR/GVK` distinction isn't handled by current code, so this change doesn't really make it any worse.

I do share the sentiment that this clientset is not good enough in the long term: we need a better way to mock/record client calls in tests. The pre-existing behavior however is very confusing, so may I suggest taking this change to fix the existing tests (and add filtering capabilities for any future tests that may need it), and plan a better overall approach for the fake clientset in the future.

caesarxuchao commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go:
```diff
@@ -111,9 +111,12 @@ func (g *genClientset) GenerateType(c *generator.Context, t *types.Type, w io.Wr
 var common = `
-// Clientset returns a clientset that will respond with the provided objects
+// NewSimpleClientset returns a clientset that will respond with the provided objects.
+// It's backed by a very simple object tracker that processes creates and updates as-is,
```
Shouldn't it be "creates, updates and deletions"? :)

olegshaldybin replied:
Yes, good point.

caesarxuchao commented:
@olegshaldybin I'm ok with using `GVK`, and I'm ok with the clarification in the comment of `NewSimpleClientset`. I haven't reviewed the code carefully though, I'll leave that to @deads2k. Thank you both.

olegshaldybin commented:
@caesarxuchao Thanks! @deads2k, note that I dind't change linear search on an `objects` map to using a map of maps of maps, as it was slightly more code and given the number of objects inn these clientsets (usually <10), O(n) should do fine. If you want I can change to use maps, but it's a little less readable that way.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
 	return nil
 }
 
+func filterByNamespaceAndName(objs []runtime.Object, ns, name string) ([]runtime.Object, error) {
```
doc the meaning of empty here.  Empty namespace is one of the worst constructs we have.  Sometimes it is none, sometimes it is all, sometimes it is any, and sometimes it is "default".

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
 	}
-	if index < 0 {
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+	if len(matchingObjs) == 0 {
```
`SetList` ought to handle empty list.  Does it not?

olegshaldybin replied:
It does, I'll remove the length-check.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
+	if err != nil {
+		return nil, err
+	}
+	if len(matchingObjs) == 0 {
```
make an error of some kind for getting more than one here.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
+}
+
+func (t *tracker) Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error) {
+	obj, err := t.scheme.New(gvk)
```
Use the scheme to figure out if you need to require a namespace or not.  This should not allow a get of a namespaced resource without a namespace.

deads2k replied:
I'm not seeing why you need this here.  Doesn't `Copy` already return it to you?

olegshaldybin replied:
You're right, that's some kind of leftover, will remove.

olegshaldybin replied:
Not sure how to use scheme to determine whether object is namespaced. `ObjectScheme` implements `ObjectCreater`, `ObjectCopier` and `ObjectTyper`, none of them seem to have any metadata about the object that we can use to check whether it's namespaced.

deads2k replied:
> Not sure how to use scheme to determine whether object is namespaced. ObjectScheme implements ObjectCreater, ObjectCopier and ObjectTyper, none of them seem to have any metadata about the object that we can use to check whether it's namespaced.

Oh, it moved.  `pkg/apimachiner/registered.Group` will give you `GroupMeta` which has a `RESTMapper`, which can give you the `RESTMapping`, which has the `RESTScope`.

olegshaldybin replied:
I added the check and it caught even more tests that don't set the namespace for namespaced resources. Will fix those and update the PR once they all pass.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
+}
+
+func (t *tracker) add(obj runtime.Object, replaceExisting bool) error {
+	if meta.IsListType(obj) {
```
How and why would I get here with a list type?

deads2k replied:
> How and why would I get here with a list type?

Oh, you're using this for initialization?  I think I'd suggest a different external method for initialization that breaks it apart and does the individual adds.

olegshaldybin replied:
Yeah, the existing code assumes Add can accept the list, I wasn't a big fan but decided to keep it. Will experiment with adding a separate `Init()` method.

olegshaldybin replied:
It seems that will be slightly more involved than I expected, since tests use both list and non-list objects for the fake clientset initialization and sometimes mix them together. Should I just change all tests to not add lists? The alternative is to keep the current behavior for now and revisit it along with further fake clientset improvements.

deads2k replied:
> It seems that will be slightly more involved than I expected, since tests use both list and non-list objects for the fake clientset initialization and sometimes mix them together. Should I just change all tests to not add lists? The alternative is to keep the current behavior for now and revisit it along with further fake clientset improvements.

I won't block the pull on this.  Fixing in the future would be desireable.  Initialization is special and I'd rather have accidents in unit tests.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered kinds for %v", obj)
+	}
+	gvk := gvks[0]
```
Seems like you should loop through and add it to all of them.  Hitting an error would fail the call and leave this instance in an unknown state, that's ok.

olegshaldybin replied:
Makes sense, will do that.

deads2k commented on pkg/client/testing/core/fixture.go:
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
-		o.types[kind] = append(o.types[kind], obj)
 	}
 
 	return nil
```
I should get a NotFound if I can't delete.

olegshaldybin replied:
Sounds good, will fix.

olegshaldybin commented:
@deads2k I've responded to your feedback, please see last 4 commits. Had to fix some tests after adding the namespace check. Also, noticed that Event objects are namespaced but controllers don't fill in namespace field when they send events, so added a bypass for Events in the namespace check. PTAL.

olegshaldybin commented:
@deads2k Just rebased so it doesn't get too stale.

deads2k commented:
lgtm

Can you squash it down to something around four commits?  Something like
 1. new reaction functions and tests for the harness
 1. generated code
 1. intent bugs in tests (looked like deployments may have been bugged?)
 1. "boring" test updates

To make it a little easier to find and track problems.

olegshaldybin commented:
@deads2k Squashed to 7 commits: object tracker modifications, generated code, commit per functional test fix + one commit for "boring" test fixes. Can probably squash functional test fixes together but it seemed to be slightly cleaner to have them fixed separately.

olegshaldybin commented:
@deads2k Had to rebase.

deads2k commented:
The queue isn't open to non-1.3 work right now.  Mention me again after its open to remind me to rescan and tag?

olegshaldybin commented:
Sounds good.

deads2k commented:
> Sounds good.

The queue is open again, but this needs a rebase.

olegshaldybin commented:
@deads2k Thanks, just rebased and fixed merge conflicts. Note that I had to further fix the deployment controller test, since controller logic has changed, and not it also updates deployment status as a part of replica set creation flow (cc @janetkuo).
